### ISSUE TITLE
Cache data is written to `userData` dir instead of `cache` dir

### ISSUE THREAD

**Issue by smyrman** [electron/electron#8124]
* Electron version: 1.3.9 (via Atom 1.12.6) and unknown (via Slack 2.3.2)
* Operating system: Linux

### Expected behavior

Files stored by Electron on behalf of electron apps, should follow the XDG Base Directory specification and place only config files in $XDG_CONFIG_HOME/<App name> (default ~/.config/). cache and data files should be placed in XDG_CACHE_HOME (default: ~/.cache) and XDG_DATA_HOME (default: ~/.local/share) respectively.

https://specifications.freedesktop.org/basedir-spec/latest/ar01s03.html

### Actual behavior

As seen in Atom and Slack on Linux, they leave non-config files in ~/.config/Atom and ~/.config/Slack/ in particular, they leave folders such as "databases", "GPUCache", "Cache", "logs", and "Local Storage". Atom mention this as an Electron issue in https://github.com/atom/atom/issues/8281 ("where electron stores it's files"). 

Could still be it's not Electron's fault that the apps store non-config files in these locations, but hopefully you know more than me here.

**Comment by kevinsawicki** [comment]
> Files stored by Electron on behalf of electron apps

Chrome provides the paths to these files, it looks like they've labeled issue as a won't fix in the past.

https://bugs.chromium.org/p/chromium/issues/detail?id=129861

**Comment by smyrman** [comment]
It seams to me though that chromium still uses XDG_CACHE_HOME correctly? Although they do store config and browser data together.

Is it possible for electron to move fodlers like GPUCache Cache to XDG_CACHE_HOME, or are those still defined by Chrome/chromium?

**Comment by kevinsawicki** [comment]
> Is it possible for Electron to move folders like GPUCache and Cache to XDG_CACHE_HOME, or are those still defined by Chrome/chromium?

Those files are completely managed by Chrome I believe, are you seeing standalone Chrome store them in a different location?

**Comment by smyrman** [comment]
I find ~/.cache/google-chrome/Default/
Cache  'Media Cache'
and:
~/.config/google-chrome/Default/GPUCache

, so actually, it just seams inconsistent...

Thanks for the quick reply!

**Comment by deepak1556** [comment]
> Is it possible for Electron to move folders like GPUCache and Cache to XDG_CACHE_HOME, or are those still defined by Chrome/chromium?

Yup its possible. Currently we set the cache path to app data path in brightray. @kevinsawicki are we ok on changing this ?

**Comment by kevinsawicki** [comment]
>  @kevinsawicki are we ok on changing this ?

Yeah, possibly, I'm not sure why the current behavior was implemented that way so I'd like to understand that. Also changing it means existing apps upgrading will start with a fresh cache right?

**Comment by smyrman** [comment]
> Also changing it means existing apps upgrading will start with a fresh cache right?
That doesn't sound dangerous to me, although I'm an outsider here... However to get rid of the old cache, perhaps you could have a move old cache directory on startup if it exists. Then you get to clean up users systems while keeping old cache at the same time, if that's desirable.

**Comment by deepak1556** [comment]
> Also changing it means existing apps upgrading will start with a fresh cache right?

Yup.

Maybe we can adopt this with next chromium update release or 2.0 ?

**Comment by paulcbetts** [comment]
I'm okay with moving cache folder, but please do not move user settings folder, this isn't worth breaking every app on the planet

**Comment by MarshallOfSound** [comment]
If we move the cache folder we will have to add internal logic to clean up the old one.  I don't want to update Electron versions and suddenly have two caches (which aren't negligible in size) on the users disk.

**Comment by baconface** [comment]
> I'm okay with moving cache folder, but please do not move user settings folder, this isn't worth breaking every app on the planet. Local Storage / Cookies et al are user data for many apps.

I agree with this completely. Even if the implementation is wrong it is not running in a broken state. So fixing the implementation will inadvertently break things. So unless you find a backwards compatible fix it should not be patched outright.

For a real life instance it would cause a headache as it would make our POS in all our store locations back to a default state that need to be reconfigured since we rely on Local Storage. And I imagine the same can be for many others.

**Comment by smyrman** [comment]
> I'm okay with moving cache folder, but please do not move user settings folder, this isn't worth breaking every app on the planet. Local Storage / Cookies et al are user data for many apps.

I think nobody is suggetting to move user settings... user cache should go to XDG_CACHE_HOME ($HOME/.cache by default)

Ideally, what you refer to as user data (including databases) should go in XDG_DATA_HOME ($HOME/.local/share by default), chrome doesn't In my opinion correctly do this, but many other Linux apps do.

User Settings are already in the correct place (XDG_CONFIG_HOME, defaulting to $HOME/.config).


Is it possible for Electron to simply add a startup hook (that can be removed in 2.0) that moves the old content, and if necessary, adds a symlink in the old location? And does it solve all the concern from e.g. @baconface?

**Comment by MarshallOfSound** [comment]
> Is it possible for Electron to simply add a startup hook (that can be removed in 2.0) that moves the old content, and if necessary, adds a symlink in the old location? And does it solve all the concern from e.g. @baconface?

This will be unpredictable and could cause issues with existing applications.  Permission issues, copy speeds and all sorts of other things could cause a working application to suddenly become non-working.  This change should be opt-in not automatic IMO.

**Comment by smyrman** [comment]
@MarshallOfSound, I just want to remind you that a move is not affected by copy speed within the same disk partition. I would also point out that it's probably not very common for users to have "$HOME/.config/" and "$HOME/.cache/" on the same disk, in which case the copy speed would be a factor.

I am not denying that it can cause issues, although it's not clear for me what those issues really are, and if they are real or not...

**Comment by MarshallOfSound** [comment]
My point is basically, moving this folder could have unpredictable side-effects for existing users so make this change opt-in until `2.0` and then make it opt-out.  Why risk it breaking when you can just stick with what you know works?

I always like to err on side of caution 👍

**Comment by smyrman** [comment]
> My point is basically, moving this folder could have unpredictable side-effects for existing users so make this change opt-in until 2.0 

You decide how you want to manage risks in your project, as long as you are happy improving on it long term 👍

**Comment by da2x** [comment]
The same problem exist on macOS as well. Cache should be stored in "~/Library/Caches" not "~/Library/Application Support".

When this is setup correctly then many backup utilities (Time Machine, Dejá Dùp) and indexing agents (Spotlight, tracker) will know not to backup or index the contents of these directories.

**Comment by anatoli26** [comment]
What do you thing about just introducing the options for the paths where to store each category of files (user specific configuration ($XDG_CONFIG_HOME), user-specific data ($XDG_DATA_HOME) and user specific non-essential data files ($XDG_CACHE_HOME)) with their default values set to the current locations? This way each app based on Electron could implement its own way of handling the issue.

I'm using 2 e-based apps: Atom editor & Wire messenger. This topic for Atom is already quite discussed in the linked issue, the Wire issue is this: https://github.com/wireapp/wire-desktop/issues/521.

What happens with Wire is that the cache stores all transferred files, among other things, and the `.config` folder grows to GBs, frequently exceeding 10Gb. For a home user this is not an issue, but for enterprise environments this is a major issue and a reason for not adopting the app.

Based on the discussion in the above-mentioned Wire issue, there's no way for Wire to define the paths on its own, the change should be implemented in Electron.

If Electron would have variables for each group of files, Wire developers could manage the migration of current date in each folder for their app. At the same time, all other apps would be unaffected. And once their developers discover the new options, they could make a decision best suited for their users.

**Comment by smyrman** [comment]
> with their default values set to the current locations? 

Not sure this makes sense. A big part of the XDG specification is it's definition for defaults. However, if there was a ELECTRON_USE_XDG the user could set, that would be a good start.

**Comment by anatoli26** [comment]
Yep, the idea is to define 3 internal variables for the 3 types of data, by default assigned to the current paths (so nobody is affected by the change), that would be configurable by electron-based apps.

Once this change is implemented and the 3 paths become available for configuration to e-apps developers, the entire process of migrating existing data of already released e-apps to the new paths, that was debated here, would be shifted to the e-apps developers, so each dev team would decide if their apps should follow the XDG specification or not. And if they decide to follow it, they could/would assign the default values to the ones mandated by the specification (it would be up to them to make the default paths decision and how to manage the migration).

So the idea is not to implement the XDG specification in electron itself, but to make it possible for the developers of e-apps to *decide* how to manage their data (today they don't have this choice).

**Comment by smyrman** [comment]
> The idea is to define 3 internal variables in Electron for the 3 types of data, by default assigned to the current paths (so nobody is affected by the change).

Alright. 

> So the idea is not to implement the XDG specification in electron itself, but to make it possible for the e-app developers to decide how to manage their data (today they don't have this choice).

It's certainly a step in the right direction.

**Comment by zcbenz** [comment]
Actually the cache directory has been set to the correct places in Electron, it is Chromium that does not respect the cache directory when storing cache data. Chromium just writes everything to `userData` directory, the logic is hard-coded everywhere and it is impossible to patch all of them.

**Comment by blahah** [comment]
@zcbenz is there any progress on this? Is it blocked at the chromium level?

**Comment by reitzig** [comment]
Any movement here? Is further discussion needed?

> So fixing the implementation will inadvertently break things.

If a vanishing caches breaks things, your app is already broken.

This is a huge pain in the ass. Electron apps (Rocket.Chat, Discord, ...) store hundreds of megabytes of cache data in `.config` which really bloats backups (unnecessarily so) and is a hassle to workaround. So, it's not only wrong, it's also actively harmful.

I get that moving user data may be dangerous -- fine. Leave it be for now, with plans to migrate to the "right" place eventually. I'll probably want to back those up, anyway. If you just move the cache folder to a suitable place (which, again, should never break anything) now. Please?

**Comment by mb21** [comment]
> My point is basically, moving this folder could have unpredictable side-effects for existing users so I would like it if this change is opt-in until 2.0 and then make it opt-out.

Sounds good... maybe this can be done for 3.0?

**Comment by MurzNN** [comment]
4.0 already released, but still no progress with this issue? Is there any blocking issues exists for implementing this as opt-in (via special configuration option) in next release?

**Comment by flokli** [comment]
I guess @MarshallOfSound already sketched how this should be implemented - honoring `XDG_CACHE_HOME` via an opt-in, if active, move old cache folder if present to avoid files lingering around - flipping it to be an opt-out for the next major release.

No need to worry about `XDG_CONFIG_HOME` and `XDG_DATA_HOME` for now.

Anybody up to make a PR doing exactly that?

**Comment by MicahZoltu** [comment]
I would like to advocate fairly strongly that Electron _default_ to respecting the operating system provided storage locations.  Whatever default Electron chooses is what _the majority of apps_ will use.  I'm looking at my hard drive right now and I have a dozen Electron apps that are all incorrectly writing a ton of cache data and program data (e.g., binaries/scripts) to `%APPDATA%` instead of the proper location of `%LOCALAPPDATA%`.  It totals out to maybe 500MB of data that should live elsewhere.

If you just provide an _option_ for Electron app developers to write elsewhere, the vast majority of them won't because they don't understand the ramifications of their choices.  You (Electron devs) are in a position to make the world suck less by providing defaults that respect the operating system recommendations, rather than providing defaults that actively go against operating system defaults.

I do recognize that migration is hard, and I will generally support any plan that _eventually_ gets the _default_ storage location for Electron apps to the right place.  What I'm actively against is kicking the can down to Electron app developers with a default that is wrong.

**Comment by rhysperry111** [comment]
I see this issue is being discussed a lot and a lot of people have been asking where it is going. Is anyone _actually_ intending on fixing this?

**Comment by MurzNN** [comment]
As workaround for this problem we can use [unburden-home-dir](https://github.com/xtaran/unburden-home-dir) script with this mask:
```
# Electron apps in .config folder
m D .config/*/Cache electronApp-%1-Cache
m D .config/*/GPUCache electronApp-%1-GPUCache
m D .config/*/*/Cache electronApp-%1-%2-Cache
m D .config/*/*/GPUCache electronApp-%1-%2-GPUCache

# custom folder locations for Electron apps in other folders
```

This will move all Electron apps caches to separate folder.

**Comment by 532910** [comment]
Is there any update?

**Comment by reitzig** [comment]
Just checking out data sinks in my home folder. This _sucks_:

![image](https://user-images.githubusercontent.com/4246780/85539432-1c62e380-b616-11ea-8c19-e5f718c6d9d1.png)

This _needs_ to be fixed. Transient data has _no business_ polluting `.config`!

NB: electron is not the only sinner of this, but the sheer amount of data dumped here is singular, at least on my machine.

**Comment by reitzig** [comment]
An alternative to consider: follow the [cachedir](https://bford.info/cachedir/) conventions.
It's almost trivial to adopt: just add a single file to all cache directories.

While Electron would then still bloat up directories it shouldn't, most if not all backup tools are able to ignore directories based on tag files.

Other use cases remain hard, though, e.g. (manually) cleaning out (all) caches. `rm -rf /tmp/* ~/.cache/*` is what you want to be able to do.

**Comment by smyrman** [comment]
@reitzig, it could be a good stop-gap, but not an acceptable final solution, although it could be part of it.

However, I feel this should really have been solved three years ago. Could I propose a 4-step fix for who ever feels comfortable doing a PR?

1. Apply @ritazh's work-around.
2. Add a global option `useOSCacheDir`(default `false`) that applications can set (or for that matter, end-users), respecting Mac OS, Linux (XDG_CACHE_HOME) and Windows defaults. 
3. For the next major release, change the default.
4. For the next major release after that, remove the option.

Leave looking at handling non-config / non-cache (XDG_DATA_HOME) to a different issue.

**Comment by reitzig** [comment]
NB: `XDG_CACHE_HOME` is -- sadly! -- _not_ a universally used standard across all Linux distributions, so a complete solution will have to consider more than that.

**Comment by 532910** [comment]
> not a universally used standard across

the same as $XDG_CONFIG_HOME

Distro is XDG compliant and has both of them, or no one at all.

**Comment by tzarebczan** [comment]
We have a couple users also complaining about the on LBRY. Please prioritize if possible :)

**Comment by haarp** [comment]
> NB: `XDG_CACHE_HOME` is -- sadly! -- _not_ a universally used standard across all Linux distributions, so a complete solution will have to consider more than that.

Not a problem, as the fallback for lack of the variable is still `$HOME/.cache`.

> As workaround for this problem we can use [unburden-home-dir](https://github.com/xtaran/unburden-home-dir) script with this mask:
> ...
> This will move all Electron apps caches to separate folder.

This sadly doesn't help. You can symlink the caches somewhere else, but apps delete the symlink and recreate the cache directories on start. Infuriating.

**Comment by zxcat** [comment]
Hey, this lasts for years, any updates?

Current defaults are just harmful and multiply bad practices in apps. macOS users forced to add more and more excluded folders in Time Machine to prevent unnecessary grow of backups. Sometimes these backed up files lead to deletion of possible useful old data. It just sucks. Spotlight search becomes broken too.

I see no problem of moving cache and fully agree with @MicahZoltu https://github.com/electron/electron/issues/8124#issuecomment-495976867 — if app can't handle cache deletion, then it just broken. So if some app breaks after fixing cache dir, it will only make the world better, so devs will fix their app to work properly.

Even if we assume that this change is really breaking (which I disagree with), this is still not a problem. Breaking changes happen anyway, and developers will adapt their applications for them. This is usual. Sooner or later, you will have to correct the default values, but the longer you wait, the more problematic apps are created that make the world worse.

Therefore, please pay attention to this issue.

**Comment by anatoli26** [comment]
Electron devs, if someone submits a PR for this issue, will it be considered?

**Comment by electron-triage** [comment]
The Electron version reported on this issue is no longer supported. See our [supported versions documentation](https://www.electronjs.org/docs/tutorial/support#supported-versions).

**If you're still experiencing this issue on a supported version, please open a new issue with an updated repro - a [Fiddle](https://www.electronjs.org/fiddle) is very appreciated.**

Electron has a large issues backlog. To help our team prioritize, we're closing older issues and asking for new issues with updated repro steps if it affects a supported version. This helps sort what issues are still relevant and helps us fix them more quickly.

Thanks for your patience and understanding!

**Comment by tjanez** [comment]
> The Electron version reported on this issue is no longer supported. See our [supported versions documentation](https://www.electronjs.org/docs/tutorial/support#supported-versions).

Huh, it is perhaps a bit unfortunate that this issue has a version number in @smyrman's original description...

But this is still a valid issue which gathered some support and subscribers so I think it would be worthwhile to re-open it and somehow prevent @electron-bot from closing it again.

@deepak1556, could you re-open it?

**Comment by tjanez** [comment]
Thanks for re-opening, @MarshallOfSound.

I haven't noticed PR https://github.com/electron/electron/pull/19693 before but I'm curious if it would resolve this issue?

**Comment by RubenKelevra** [comment]
Can please someone fix this? It's super annoying to add all the exclusions on backup programs. We have a cache folder for this, writing stuff to .config except configs should not happen.

**Comment by idontusenumbers** [comment]
I just cleared tens (maybe even hundreds) of gigabytes from my MacOS TimeMachine backup by laboriously going through my backups using commercial software to identify large backups, finding large change sets, invoking (buggy) time machine features to delete these cache folders for numerous Electron apps on my system.

These caches last forever in the weekly backups going back (in my case) to 2016.

**Comment by nornagon** [comment]
I think this is a duplicate of #1404?

**Comment by tjanez** [comment]
> I think this is a duplicate of #1404 (or vice-versa)?

I haven't read that issue, but from a quick glance I would say this issue has a lot more participants and upvotes, so please don't close it :slightly_smiling_face:.

**Comment by MurzNN** [comment]
Here is fresh PR with fixing this issue: https://github.com/electron/electron/pull/30665 - hope it lands in near future!

**Comment by DC7IA** [comment]
> hope it lands in near future!

It didn't :thinking: 

Hope it will land at all at some point. :)

**Comment by tutunak** [comment]
Sad to hear that this issue still opened.

**Comment by MurzNN** [comment]
No-no-no! Please fix this!

**Comment by MicahZoltu** [comment]
Ugh, this is still definitely a problem and it just gets worse the more things that get built using Electron.  Please do not close this!

**Comment by admincheg** [comment]
6 years and still no solution? Are you serious electron?

In my case I have a lot of RAM and place all cache into it, because I like my NVME disks and do not want spend their health for temporary data. I've already dropped off chrome/chromium, but some applications can not be dropped, sad, but true.

Just as mentioned before - Linux has a kind of standard - XDG_CACHE_HOME environment variable. Why you still ignore it, guys?

**Comment by RobinDaugherty** [comment]
Mac OS and backup software know how to treat cache and other high-turnover low-importance data, but Electron ignores the Mac OS conventions. This is just about the best example of Electron causing the industry to make worse software. Application developers expect Electron to provide a layer that allows their app to be a good player on each platform.

But on Mac, the most popular platform for developers, Electron apps are storing caches in a place that causes excessive SSD wear and backup size bloat.

This is costing people real money. I pay for storage of my online backups, and with each Electron app, more and more of my money is being spent to _back up cache files_!

**Comment by metov** [comment]
Just wanted to add my +1 to this. The XDG standard has sensible reasons for recommending the directories it does, and this has many benefits. The vast majority of Unix software already follows it, users are accustomed to expecting it, and many useful tools exist that assume it is the case.

Many programs use Electron, and all inherit this bug. It is not practical for N developers of N projects to add their own workaround N times. It would be much more efficient if Electron simply used the correct directories to begin with.

**Comment by danisztls** [comment]
> > Is it possible for Electron to simply add a startup hook (that can be removed in 2.0) that moves the old content, and if necessary, adds a symlink in the old location? And does it solve all the concern from e.g. @baconface?
> 
> This will be unpredictable and could cause issues with existing applications. Permission issues, copy speeds and all sorts of other things could cause a working application to suddenly become non-working. This change should be opt-in not automatic IMO.

Yet I'm doing this for two years with 0 issues.

**Comment by idontusenumbers** [comment]
> This will be unpredictable and could cause issues with existing applications.

The existing behavior, right now, 100% always causes problems for anyone using industry-standard backup practices. Backing up cache directories should be the thing people need to opt-in to. If the copy is the problem or takes longer than some arbitrary period, just obliterate the old data; it's a cache directory, there's no expectation that it sticks around.

**Comment by Volatus** [comment]
Can't Electron simply check if the old directories already exist and if so use them for backwards compatibility while giving higher priority to XDG suggested directories?

**Comment by dontlaugh** [comment]
This is definitely going to make my rsync script more complicated, folks.

**Comment by heroin-moose** [comment]
Come on people, this is insane.

**Comment by MurzNN** [comment]
> @dontlaugh Try this https://github.com/danisztls/ephemeral

This is a really interesting and promising project, thanks for the link! But the main message of this project to use `$XDG_RUNTIME_DIR` for cache files is totally wrong, so I created issue to fix this: https://github.com/danisztls/ephemeral/issues/2

**Comment by Volatus** [comment]
I was trying to make a PR for this but the electron dev environment setup is pretty annoying to get working...

**Comment by zcbenz** [comment]
https://github.com/electron/electron/pull/34337 provides a solution which at least goes one step forward towards solving this issue, which requires review from @electron/wg-api .

**Comment by Matthieu-LAURENT39** [comment]
It's pretty crazy that this still isn't fixed, it makes backing up ~/.config a nightmare

**Comment by heroin-moose** [comment]
Gentle bump. Given that we've established that it is possible to do this in a backward-compatible manner, please, do the right thing and act on this issue. Good defaults matter.

**Comment by regalialong** [comment]
(sigh)

**Comment by nre** [comment]
> Can't Electron simply check if the old directories already exist and if so use them for backwards compatibility while giving higher priority to XDG suggested directories?

bump

**Comment by RobinDaugherty** [comment]
The response to this is not surprising, even if it is disappointing. The entire purpose of Electron seems to be to ruin desktop apps in exchange for lowering the cost of developing proofs-of-concept. Filling your RAM and eating your CPU with multiple copies of a web browser engine, ruining accessibility, and rendering automation tools useless were bad enough. Wearing out our flash storage and backup processes with log files and caches in nonstandard places is just another poke in the eye. The Electron team does not care.

**Comment by gdevenyi** [comment]
Don't forget that this causes unneeded traffic and disk-usage on NFS-based $HOME. I've had to ban Chrome here because of that.

**Comment by heroin-moose** [comment]
@codebytere, @ckerr, can you please take a look? This looks rather important to opensource community.

**Comment by deviantintegral** [comment]
At this point, I wonder if a better tactic is to file support and bug requests with software built on Electron like Slack, Discord, etc, especially if you're a paying customer. I'll be filing a Slack request from our org myself.

**Comment by 532910** [comment]
stale fact is not a reason to close issues

**Comment by RubenKelevra** [comment]
True. Just because the team seems to not care, issues don't go away.

**Comment by RubenKelevra** [comment]
How.

**Comment by RubenKelevra** [comment]
... many...

**Comment by RubenKelevra** [comment]
...replies...

**Comment by RubenKelevra** [comment]
... does...

**Comment by RubenKelevra** [comment]
... this...

**Comment by RubenKelevra** [comment]
...bot...

**Comment by RubenKelevra** [comment]
... need?

**Comment by MurzNN** [comment]
This is still reproducible in the [latest version of Electron](https://www.electronjs.org/releases/stable) or in the [beta](https://www.electronjs.org/releases/beta) and very annoying! 

The issue has 167+ upvotes! So, @electron-triage - please mark this issue to exclude it from the @electron-bot bot stale procedure!

**Comment by regalialong** [comment]
This sucks.

From the dev perspective, users have to keep hammering the issue with essentially beg comments, which adds noise to the actual important discussion above on how to implement this without breaking things (mind you, Github has already collapsed a lot of that from the amount of comments).

From the user perspective, I get notifications that aren't actually related to this being implemented.

Please add the `stale-exempt` tag to this issue, so that my notifications are for when something actually happens.

**Comment by admincheg** [comment]
Interesting question - if developers do not want to do something - why someone does not just fork this project and make it better? I think fresh-new fork repo with feature will be more popular, than deaf developers.

inb4: because nobody wants responsibility

**Comment by heroin-moose** [comment]
Oh come on. We have a solution, just fix it :DDDD

**Comment by regalialong** [comment]
> Interesting question - if developers do not want to do something - why someone does not just fork this project and make it better? I think fresh-new fork repo with feature will be more popular, than deaf developers.
> 
> inb4: because nobody wants responsibility

I don't have the time or the expertise to develop an Electron fork.
Even if I did, it likely wouldn't have widespread adoption like a proper fix would (which means in the end, applications from vendors would still end up using the current bad defaults).

**Comment by severindupouy** [comment]
Is it so complicated for a cross-platform application to properly handle these kinds of issues?

**Comment by Atrate** [comment]
I second this. Moving cache to the proper directory would also *significantly* boost performance (at least warm-cache performance) for users that keep their XDG_CACHE_DIR on `tmpfs` or an equivalent ramdisk. It would also help decrease the number of SSD writes, reducing e-waste :).

**Comment by gdevenyi** [comment]
This is an important optimization for NFS $HOME to reduce network traffic and for caching to do what its supposed to (make things local and fast)

**Comment by lhanson** [comment]
> Is it so complicated for a cross-platform application to properly handle these kinds of issues?

Narrator: It is not

**Comment by heroin-moose** [comment]
@ckerr, can you please take a look at this?

**Comment by severindupouy** [comment]
Just heard about https://github.com/lunacookies/etcetera… maybe useful.

**Comment by MurzNN** [comment]
Bump! Come on, please fix this and clean up millions of disks from storing temporary garbage!

**Comment by heroin-moose** [comment]
At this point I believe that they are simply want us to suffer.

**Comment by heroin-moose** [comment]
Guys, please, take a look. This issue has a simple solution. 

@codebytere @erickzhao @cker @deepak1556 @derekcicerone @VerteDinde @dsanders11

**Comment by regalialong** [comment]
Can you believe it guys? Bump in a week, whoohoo. I'm so happy about this information.

**Comment by gdevenyi** [comment]
Amazing.

**Comment by regalialong** [comment]
Holding users hostage to constantly bump (and get notifications for bumps on) issues that still aren't fixed is incredibly friendly and productive issue triage by the way.

**Comment by sedrubal** [comment]
This issue still affects me.

**Comment by anaisbetts** [comment]
This bug should honestly be closed as WONTFIX. While OP is probably right, the time to have fixed it is 10+ years ago; it is Too Late now. Leaving this in the hands of developers to deal with would almost certainly be a complete disaster and even trying to auto-migrate everything behind the scenes would almost certainly break applications.

**Comment by MicahZoltu** [comment]
> it is Too Late now.

Why is it too late now?  I can't think of a reason why this couldn't be fixed now but could have been fixed in the past.  It is still a bug, it is still causing problems for users, and it still has the same route to fixing it as before.

What sorts of disasters are you imagining?  What do you believe will go wrong if this is fixed?

At a bare minimum, you could first check the new correct location, and if that is empty check the old incorrect location and do not do any migration.  If both are empty, use the new location.  If old exists and new doesn't exist, continue using the old path.  This would at *least* allow for manual migration by apps and users, without forcing a migration on anyone.

**Comment by odinho** [comment]
Agree with the above. This can very well be fixed now.

We're building a new app on Electron and it'd be very nice to have this correct from the start. I'd also personally fix this for other apps too, since I do remote backup of some of these folders.

**Comment by severindupouy** [comment]
bump

**Comment by heroin-moose** [comment]
At this point Electron is beyond repair. Just use Qt guys. It’s hopeless.

**Comment by likelyrand** [comment]
almost 9 years and still not fixed, what a joke

**Comment by genodeftest** [comment]
If you don't have anything productive to add (like a PR fixing this issue), please stop it. Ranting doesn't help anyone and won't make it more likely that the devs will fix it.

**Comment by heroin-moose** [comment]
Everything productive already here. Multiple people suggested backward-compatible algorithm.

**Comment by Half-and-Xat** [comment]
> This bug should honestly be closed as WONTFIX. While OP is probably right, the time to have fixed it is 10+ years ago; it is Too Late now. 

[Mozilla fixed their issue after 21 years](https://www.phoronix.com/news/Firefox-147-XDG-Base-Directory), just a few days ago. So why electron shouldn't be able to implement this? It is just a matter of willingness. 

Today (btw issue also became 9 years old) this bug broke my automated backups and while as experienced user I can fix it, it costs time to do so. It also spams my backups with with new useless data on daily basis on programs that I am not aware of they're storing cache in `.configs` folder or similar. The one single program I realized today doing it caused 750MB (cache) backup trash in few weeks. I may should look what other installed applications are using electron. Such things creates more troubles to inexperienced users which may do not understand what logs are telling. And in additional it creates additional work to other projects, because they are getting annoying bug reports as mine which you can see above, following the link. 

People will have to deal with it (programmers **and** users) until it is fixed. So even if it takes another 12 years to fix it, it is still worth doing so.

**Comment by bysiber** [comment]
This is still a significant issue in Electron apps writing cache data to `userData` means standard cache cleanup tools miss them entirely. On macOS, this is particularly problematic because Electron apps scatter data across multiple locations:

**Quick audit of Electron app cache on macOS:**
```bash
# Electron apps in Application Support (treating cache as user data)
du -sh ~/Library/Application\ Support/Slack 2>/dev/null
du -sh ~/Library/Application\ Support/discord 2>/dev/null
du -sh ~/Library/Application\ Support/Code 2>/dev/null  # VS Code
du -sh ~/Library/Application\ Support/Microsoft\ Teams* 2>/dev/null

# Where cache SHOULD be (~/Library/Caches)
du -sh ~/Library/Caches/com.slack.Slack 2>/dev/null
du -sh ~/Library/Caches/com.microsoft.VSCode 2>/dev/null

# GPU shader cache (often overlooked)
du -sh ~/Library/Application\ Support/*/GPUCache 2>/dev/null
du -sh ~/Library/Application\ Support/*/Cache 2>/dev/null
du -sh ~/Library/Application\ Support/*/CachedData 2>/dev/null
```

On my machine, VS Code alone has ~2GB in Application Support (CachedData, CachedExtensions, User/workspaceStorage). Slack: ~1.5GB. Discord: ~800MB. This data should survive a cache clear but currently it is all mixed together with actual config.

**What you can safely clean (Electron apps on macOS):**
```bash
# These subdirectories within Application Support are cache-like and safe to remove:
rm -rf ~/Library/Application\ Support/*/GPUCache
rm -rf ~/Library/Application\ Support/*/Cache
rm -rf ~/Library/Application\ Support/*/Code\ Cache
rm -rf ~/Library/Application\ Support/*/CachedData
rm -rf ~/Library/Application\ Support/*/Service\ Worker/CacheStorage
```

For anyone wanting a visual audit of all these scattered caches, I maintain [ClearDisk](https://github.com/bysiber/cleardisk) -- a free macOS menu bar app that scans developer caches (including Electron/VS Code cache directories, Docker, Xcode, node_modules, Homebrew, etc.) and shows exactly what is using disk space. Helpful for catching these "hidden" caches that live outside `~/Library/Caches`.

**Comment by regalialong** [comment]
> 2025

Hey, you might wanna check the synchrony of your clocks (i mean both your system and internal one).

**Comment by MurzNN** [comment]
Okay, if Electron developers still ignore this critical issue and don't want to fix this, maybe we can start a community-based project with just a script that scans the `userData` directory and makes symlinks to the `XDG_CACHE_HOME` directory? 

Or maybe there is already something like this exist? If so, would be great to update the issue description with the section "Workarounds", because a lot of comments in this issue got hidden by default by GitHub.

**Comment by shvchk** [comment]
@MurzNN I've been using [danisztls/ephemeral](https://github.com/danisztls/ephemeral) for a year or two without issues
